src/java.base/share/classes/sun/security/pkcs12/MacData.java

```diff
@@ -70,103 +94,201 @@ class MacData {
+                    "(" + m.getAlgorithm() + " iterations: "
+                    + this.iterations + ")");
+        }
+
+        if (!MessageDigest.isEqual(this.digest, macResult)) {
+            throw new UnrecoverableKeyException("Failed PKCS12" +
+                    " integrity checking");
         }
     }
 
-    MacData(String algName, byte[] digest, byte[] salt, int iterations)
-        throws NoSuchAlgorithmException
-    {
-        if (algName == null)
-           throw new NullPointerException("the algName parameter " +
-                                               "must be non-null");
-
-        AlgorithmId algid = AlgorithmId.get(algName);
-        this.digestAlgorithmName = algid.getName();
-        this.digestAlgorithmParams = algid.getParameters();
-
-        if (digest == null) {
-            throw new NullPointerException("the digest " +
-                                           "parameter must be non-null");
-        } else if (digest.length == 0) {
-            throw new IllegalArgumentException("the digest " +
-                                                "parameter must not be empty");
+    /*
+     * Calculate MAC using HMAC algorithm (required for password integrity)
+     *
+     * Hash-based MAC algorithm combines secret key with message digest to
+     * create a message authentication code (MAC)
+     */
+    static byte[] calculateMac(char[] passwd, byte[] data,
+            String macAlgorithm, int macIterationCount, byte[] salt)
+            throws IOException, NoSuchAlgorithmException {
+        final byte[] mData;
+        final PBEParameterSpec params;
+        String algName = "PBMAC1";
+        String kdfHmac;
+        String hmac;
+        Mac m;
+        int keyLength;
+
+        macAlgorithm = macAlgorithm.toLowerCase(ENGLISH);
+        if (macAlgorithm.startsWith("pbewith")) {
+            kdfHmac = MacData.parseKdfHmac(macAlgorithm);
+            hmac = MacData.parseHmac(macAlgorithm);
+            if (hmac == null) {
+                hmac = kdfHmac;
+            }
+        } else if (macAlgorithm.startsWith("hmacpbe")) {
+            algName = macAlgorithm.substring(7);
+            kdfHmac = macAlgorithm;
+            hmac = macAlgorithm;
         } else {
-            this.digest = digest.clone();
+            throw new ParsingException("unexpected algorithm '"
+                    +macAlgorithm+ "'");
         }
 
-        this.macSalt = salt;
-        this.iterations = iterations;
+        params = new PBEParameterSpec(salt, macIterationCount);
 
-        // delay the generation of ASN.1 encoding until
-        // getEncoded() is called
-        this.encoded = null;
+        try {
+            m = getMac(macAlgorithm, passwd, params, data, kdfHmac, hmac, 0);
+            byte[] macResult = m.doFinal();
 
+            DerOutputStream bytes = new DerOutputStream();
+            bytes.write(encode(algName, macResult, params, kdfHmac, hmac,
+                    m.getMacLength()));
```

Comment:
`m.getLength()` is the same as `macResult`.

Reply:
fixed